Fixes #7958 - do not remove Battle Gear equipment when changing class

cmajames opened the pull request:
#7958
### Changes

website/client-old/js/controllers/userCtrl.js#38: removed to keep equipt equipment constant
website/common/common/script/ops/changeClass.js#33:  removed to stop 'classes' introduction

Commit:
Changed files to fix Bug 7958:
 - website/client-old/.../userCtrl.js#38: removed to keep inventory constant
 - website/common/.../changeClass.js#33:  removed to stop 'classes' introduction

Commit:
Changed files to fix Bug 7958:
 - website/client-old/.../userCtrl.js#38: removed to keep inventory constant
 - website/common/.../changeClass.js#33:  removed to stop 'classes' introduction

Commit:
Changed files to fix Bug 7958:
 - website/client-old/.../userCtrl.js#38: removed to keep inventory constant
 - website/common/.../changeClass.js#33:  removed to stop 'classes' introduction

crookedneighbor commented on website/common/script/ops/changeClass.js:
```diff
@@ -35,8 +35,6 @@ module.exports = function changeClass (user, req = {}, analytics) {
         }
       }
 
-      user.items.gear.equipped[type] = foundKey;
```
Looks like all of this `foundKey` logic can be removed as it's only used for this line.

You'll need to update the corresponding test as well.

crookedneighbor commented:
What's the status on this PR?

Commit:
Adjustments following Bug Review
 - Removed remaining 'foundKey' logic
 - Adjusted test logic to reflect feature change

Commit:
Adjustments following Bug Review
 - Removed remaining 'foundKey' logic
 - Adjusted test logic to reflect feature change

cmajames commented:
@crookedneighbor Sorry that I haven't been able to follow up - I run a tight schedule.  I've gone ahead and pushed the changes you suggested.

crookedneighbor reviewed (dismissed):
@Alys was your intention to have the class tour stuff removed entirely (even when first achieving level 10)? Because that's what happens in this PR. I would think you should go through the tour when you first choose a class, but not after.

Alys commented:
The intention is to never remove the gear that the player is wearing when they change class, not even on the first class change (because they are almost certain to be wearing better gear than the default gear for the new class; also keeping their current gear equipped when they change class for the first time emphasises that the gear is still usable).

If the tour is kept for the first class change, it will need to be modified since it currently says that your gear has been removed. Once this change is in place, the only useful part of the tour is "Wearing your class's gear grants you a 50% bonus to stats" which is also stated (more or less) when you hover over class gear. I'm not opposed to us having a different class tour screen for the first change.

Commit:
Reverting userCtrl.js to development version
 - Reintroduces "classes" Guide tour

Commit:
Reverting userCtrl.js to development version
 - Reintroduces "classes" Guide tour

cmajames commented:
@crookedneighbor is there a reason why userCtrl#changeClass handles both the prompting dialogue and actually setting the new class? If not, what sort of terrible things would happen if I wanted to refactor that segment of code?

crookedneighbor commented:
Go ahead and refactor it.

Commit:
New version of Fixes #7958
 - Changed logic to only notify user the first time they choose a class
 - Changed message to represent this change in logic
 - #LINT: Cleaned interface for changing class
    - New method: enableClasses() -- because, really, should we be calling User.changeClass({}) from the UX?
    - New method: payForNewClass() -- handles prompting the user to confirm that they want to change class

Commit:
New version of Fixes #7958
 - Changed logic to only notify user the first time they choose a class
 - Changed message to represent this change in logic
 - #LINT: Cleaned interface for changing class
    - New method: enableClasses() -- because, really, should we be calling User.changeClass({}) from the UX?
    - New method: payForNewClass() -- handles prompting the user to confirm that they want to change class

cmajames commented:
The test for api-v3 fails, however no tests are shown as being invalid; rather there appears to be an issue with the testing harness.

Alys commented:
@cmajames I've restarted the Travis CI build for your PR, so check back in a bit to see if it was only a temporary problem.

cmajames commented:
@Alys, yep, it appears the bug was some weird fluke.

cmajames commented:
@crookedneighbor I'd like to hear your insight on the current state of things if you have the time to take a look at things.

crookedneighbor commented on website/server/models/user/schema.js:
```diff
@@ -233,6 +233,7 @@ let schema = new Schema({
     armoireEmpty: {type: Boolean, default: false},
     cardReceived: {type: Boolean, default: false},
     warnedLowHealth: {type: Boolean, default: false},
+    shownClassIntro: {type: Boolean, defalut: false},
```
Not a fan of adding another user flag. 😞

cmajames replied:
"Unfortunate But Necessary"? Or can you think of another solution that doesn't involve adding flags?

cmajames replied:
User.user.flags.tour.classes?

Commit:
Remove new User Flag, use flags.tour.classes

Commit:
Remove new User Flag, use flags.tour.classes

cmajames commented on website/client-old/js/services/guideServices.js:
```diff
@@ -64,11 +64,15 @@ function($rootScope, User, $timeout, $state, Analytics) {
-        }, {
+          hideNavigation: true
+        }
+        /*, {
```
@crookedneighbor Do you know if any of the following tour elements are still used in the code?

cmajames replied:
As far as I can tell, no; a quick git grep only shows the one call to 'classes' in userCtrl. See [Gist](https://gist.github.com/cmajames/91d9469ed080589ebeba004f7c78b0e6).

TheHollidayInn replied:
I think not since it looks like this was already commented out.

Commit:
Whoopsie. Fix PR conflict.

Commit:
Whoopsie. Fix PR conflict.

TheHollidayInn commented on website/client-old/js/controllers/userCtrl.js:
```diff
@@ -21,16 +21,17 @@ habitrpg.controller("UserCtrl", ['$rootScope', '$scope', '$location', 'User', '$
+      return User.changeClass({});
+    }
+
+    $scope.payForNewClass = function(){
```
Why is this called `payForNewClass?`

TheHollidayInn commented on website/server/controllers/api-v3/user.js:
```diff
@@ -77,6 +77,7 @@ let updatablePaths = [
   'flags.cardReceived',
   'flags.warnedLowHealth',
   'flags.newStuff',
+  'flags.shownClassIntro',
```
Where is this flag used?

cmajames replied:
It isn't; I'll squash.

TheHollidayInn commented:
Everything here looks good to me and works locally. I just had a question about the flag, but I think we can merge this in.

Commit:
Removed Extraneous Flag

Commit:
Merge branch 'bug-7958' of github.com:cmajames/habitrpg into bug-7958

Commit:
Removed Extraneous Flag

Commit:
New version of Fixes #7958
 - Changed logic to only notify user the first time they choose a class
 - Changed message to represent this change in logic
 - #LINT: Cleaned interface for changing class
    - New method: enableClasses() -- because, really, should we be calling User.changeClass({}) from the UX?
    - New method: payForNewClass() -- handles prompting the user to confirm that they want to change class

Remove new User Flag, use flags.tour.classes

Whoopsie. Fix PR conflict.

Removed Extraneous Flag

Commit:
Merge branch 'develop' of https://github.com/HabitRPG/habitica into bug-7958

cmajames commented:
I put the "flake" in "special snowflake" ;)
I've purged the last remnants of the additional flag code and rebased onto the latest code as cleanly as I could ^^;

Commit:
Fixes handling architecture change

Alys commented on test/common/ops/changeClass.js:
```diff
@@ -70,13 +70,9 @@ describe('shared.ops.changeClass', () => {
       expect(user.items.gear.owned.armor_rogue_1).to.be.true;
-      expect(user.items.gear.equipped.shield).to.equal('shield_rogue_0');
       expect(user.items.gear.owned.shield_rogue_0).to.be.true;
-      expect(user.items.gear.equipped.head).to.equal('head_base_0');
```
You've removed the old tests which showed that the base level equipment of the new class was equipped, which is great, but can you please also add tests to show that the player's original choice of gear was not unequipped? For example, set the test account to own some mid-level Healer items and be wearing them as Battle Gear, then after the test does the class change, check that they're still equipped.

Alys commented on website/common/locales/en/npc.json:
```diff
@@ -61,7 +61,7 @@
 
     "classGear": "Class Gear",
-    "classGearText": "First: don't panic! Your old gear is in your inventory, and you're now wearing the apprentice equipment of your new class. Wearing your class's gear grants you a 50% bonus to stats. However, feel free to switch back to your old gear.",
+    "classGearText": "Congratulations on choosing a class! I've gone ahead and added your new basic weapon to your inventory. Take a look below to equip it!",
```
Can you remove "gone ahead and"? The sentence still makes sense without it and keeping it short is good.

Alys replied:
A question for the other admins:

The text in `classGearText` now says almost the opposite of what it used to say. If we keep using that key, anyone using a language other than `en` will get the wrong message until the translators update it, and for some languages that could be months. What's our usual process in that situation? Should we use a new key for this new text and delete the old one?

paglias replied:
When a new english string is uploaded to Transifex all old translations are removed, so there should be no problems

Alys commented on website/common/script/ops/changeClass.js:
```diff
@@ -44,26 +43,6 @@ module.exports = function changeClass (user, req = {}, analytics) {
     user.flags.classSelected = true;
 
     each(['weapon', 'armor', 'shield', 'head'], (type) => {
```
`armor` and `head` can be removed from this `each`

It might even be clearer to get rid of the `each` and just have lines like below, but I'm not certain this is better so disagreements are welcome. :)

    user.items.gear.owned[`weapon_${klass}_0`] = true;
    if (klass === 'rogue')  user.items.gear.owned[`shield_${klass}_0`] = true;

paglias replied:
agreed the code now only runs for `(type === 'weapon' || (type === 'shield' && klass === 'rogue')`

Alys commented on test/common/ops/changeClass.js:
```diff
@@ -70,13 +70,9 @@ describe('shared.ops.changeClass', () => {
-      expect(user.items.gear.equipped.weapon).to.equal('weapon_rogue_0');
       expect(user.items.gear.owned.weapon_rogue_0).to.be.true;
-      expect(user.items.gear.equipped.armor).to.equal('armor_rogue_1');
       expect(user.items.gear.owned.armor_rogue_1).to.be.true;
```
I think this `expect` for armor_rogue_1` could be removed, along with `user.items.gear.owned.armor_rogue_1 = true;` above. I believe their only purpose was to test that the armor was equipped when it was already owned, which is no longer relevant.

Alys reviewed (dismissed):
@cmajames Sorry it's taken us a while to look at this. :(

Note to other admins: I haven't done a thorough code review.
I've done some testing and it looks good so far.

Commit:
Merge remote-tracking branch 'upstream/develop' into bug-7958

Commit:
Merge remote-tracking branch 'upstream/develop' into bug-7958

Commit:
Updates following Review 20170418-0602

Commit:
Remove cause of mocha/no-exclusive-tests lint failure

cmajames commented:
Okay, that should handle all the requested changes that I've seen :)

SabreCat commented:
The fix is working, but the equipment-related notification shows up with the wrong timing now:

![image](https://user-images.githubusercontent.com/4501321/26843791-3e16e6e0-4aa6-11e7-8867-c73d6fc3e42d.png)

I think this is related to the changes made in the client `userCtrl.js`?

SabreCat reviewed (changes requested):
Need to address timing of the Justin popover, then this will be ready methinks!

SabreCat committed:
Merge branch 'develop' into bug-7958

SabreCat commented:
Upon revisiting, I am no longer able to reproduce that issue. `¯\_(ツ)_/¯` So this is good to go! @cmajames, what's your Habitica UUID from https://habitica.com/#/options/settings/api ? I can get you Blacksmith credit for this!
